pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -0,0 +1,534 @@
+    final String offline = "OFFLINE";
+
+    if (tableTypeStr == null) {
+      ret.put(formatSegments(tableName, CommonConstants.Helix.TableType.valueOf(offline)));
+      ret.put(formatSegments(tableName, CommonConstants.Helix.TableType.valueOf(realtime)));
+    } else {
+      ret.put(formatSegments(tableName, CommonConstants.Helix.TableType.valueOf(tableTypeStr)));
+    }
+    return ret.toString();
+  }
+
+
+  @GET
+  @Produces(MediaType.APPLICATION_OCTET_STREAM)
+  @Path("/segments/{tableName}/{segmentName}")
+  @ApiOperation(value = "Download a segment", notes = "Download a segment")
+  public Response downloadSegment(
+      @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName,
+      @ApiParam(value = "Name of the segment", required = true) @PathParam("segmentName") @Encoded String segmentName,
+      @ApiParam(value = "realtime|offline", required = false) @QueryParam("type") String tableTypeStr
+  ) {
+    FileUploadPathProvider provider;
+    try {
+      provider = new FileUploadPathProvider(_controllerConf);
+    } catch (Exception e) {
+      throw new WebApplicationException(e, Response.Status.INTERNAL_SERVER_ERROR);
+    }
+    try {
+      segmentName = URLDecoder.decode(segmentName, "UTF-8");
+    } catch (UnsupportedEncodingException e) {
+      String errStr = "Could not decode segment name '" + segmentName + "'";
+      LOGGER.info(errStr);
+      throw new WebApplicationException(errStr, Response.Status.BAD_REQUEST);
+    }
+    final File dataFile = new File(provider.getBaseDataDir(), StringUtil.join("/", tableName, segmentName));
+    if (!dataFile.exists()) {
+      throw new WebApplicationException("Segment " + segmentName + " or table " + tableName + " not found",
+          Response.Status.NOT_FOUND);
+    }
+    Response.ResponseBuilder builder = Response.ok(dataFile);
+    builder.header("Content-Disposition", "attachment; filename=" + dataFile.getName());
+    return builder.build();
+  }
+
+  @DELETE
+  @Produces(MediaType.APPLICATION_JSON)
+  @Path("/segments/{tableName}/{segmentName}")
+  @ApiOperation(value = "Deletes a segment", notes = "Deletes a segment")
+  public SuccessResponse deleteOneSegment(
+      @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName,
+      @ApiParam(value = "Name of the segment", required = true) @PathParam("segmentName") @Encoded String segmentName,
+      @ApiParam(value = "realtime|offline", required = true) @QueryParam("type") String tableTypeStr
+  ) {
+    CommonConstants.Helix.TableType tableType = Constants.validateTableType(tableTypeStr);
+    if (tableType == null) {
+      throw new WebApplicationException("Table type must not be null", Response.Status.BAD_REQUEST);
+    }
+    try {
+      segmentName = URLDecoder.decode(segmentName, "UTF-8");
+    } catch (UnsupportedEncodingException e) {
+      String errStr = "Could not decode segment name '" + segmentName + "'";
```

Comment:
Yes, I wanted the segment name in the error string. I don't believe we have APIs that scan for this specific error string return.